### PR TITLE
Parse search parts and conjunct resulting filters

### LINKED ISSUE #393: Implement simple search parser
Our current search strategy supports one feature at a time: keyword search, owner search, Chrome version search, etc. I propose that we should support these together using the following strategy:

1. Construct quoted and/or space-separated "atoms"
2. Merge atoms with special meaning: e.g., `owner : rbyers` becomes _owner-property-contains-"rbyers"_
3. Conjunct filters of resulting atoms; e.g., _title-or-description-contains-"dom"_ `AND` _shipped-milestone-property-greater-than-30_ `AND` _owner-property-contains-"rbyers"_

The most common use case here will probably just be finding descriptions that contain "keyword1" and "keyword2" from the search string: `keyword1 keyword2`, rather than interpreting the whole thing as a regular expression or string. That said, I'm confident that some users will find combining syntactic elements useful too.

### PR DESCRIPTION
**Pull request by mdittmer**
Merge after #392. Closes #393.

### CODE REVIEW

File: static/elements/chromedash-legend.html [hunk 1]
```diff
@@ -42,14 +42,11 @@
...
       <ul class="queries">
-        <li><span>"&lt;30"</span>features that landed before 30</li>
-        <li><span>"&lt;=30"</span>features in 30</li>
-        <li><span>"=30"</span>features that landed in 30</li>
-        <li><span>"&gt;28"</span>features since 28</li>
-        <li><span>"behind a flag"</span>all experimental features</li>
-        <li><span>"origin trial"</span>all features available as origin trials</li>
-        <!-- <li><span>"category: CSS"</span>features in the category CSS</li> -->
-        <li><span>"owner: user@example.org"</span>features owned by user@example.org</li>
+        <li><span>"shipped_milestone&lt;30"</span>features that landed before 30</li>
```
Comment (mdittmer) [inline static/elements/chromedash-legend.html]: Notice change for versioning expressions. They use _<property_name>_ _<operator>_ _<integer>_

File: static/elements/chromedash-featurelist.html [hunk 2]
```diff
@@ -72,7 +72,7 @@
           type: Boolean,
           readOnly: true,
           value: true
-        }
+        },
```
Comment (ebidel) [inline static/elements/chromedash-featurelist.html]: hanging ","!

File: static/elements/chromedash-featurelist.html [hunk 3]
```diff
@@ -130,38 +178,72 @@
...
+            } else if (match[2] && match[2].match(/[<>=]/) !== null &&
+                       i < reMatches.length - 1) {
+              operatorQueries.push([match[1], match[2], reMatches[i + 1][1]]);
+              i++;
+            } else {
+              keywordQueries.push(match[1]);
+            }
+          }
+
+          var filters = propertyQueries.map(
```
Comment (ebidel) [inline static/elements/chromedash-featurelist.html]: This is hard to grok. Can you add comments and maybe have separate lines with good var descriptions?

File: static/elements/chromedash-featurelist.html [hunk 4]
```diff
@@ -130,38 +178,72 @@
...
+          while (match = regExp.exec(val) !== null) {
+            reMatches.push(match);
+          }
+
+          // Accumulate keyword and property queries.
+          var propertyQueries = [];
+          var operatorQueries = [];
+          var keywordQueries = [];
+          var i;
+          for (i = 0; i < reMatches.length; i++) {
```
Comment (ebidel) [inline static/elements/chromedash-featurelist.html]: `for (var i = 0; i...`

File: static/elements/chromedash-featurelist.html [hunk 5]
```diff
@@ -130,38 +178,72 @@
...
+            match = reMatches[i];
+            if (match[2] && match[2].match(':') !== null &&
+                i < reMatches.length - 1) {
+              propertyQueries.push([match[1], reMatches[i + 1][1]]);
```
Comment (ebidel) [inline static/elements/chromedash-featurelist.html]: some examples would go a long way :)

File: static/elements/chromedash-featurelist.html [hunk 6]
```diff
@@ -130,38 +178,72 @@
...
           } else {
-            // Only one of the platforms needs to match.
-            // Do a general search over misc fields.
-            // Remove "=" prefix so "in development"/"proposed" version queries
-            // return results.
-            if (val.indexOf('=') == 0) {
-              val = val.substring(1);
+            var results = this.features;
+            var j;
+            for (i = 0; i < filters.length; i++) {
```
Comment (ebidel) [inline static/elements/chromedash-featurelist.html]: i -> j?

`for (var i = 0;...`

File: static/elements/chromedash-featurelist.html (hunk 4 above)
Comment (mdittmer) [inline static/elements/chromedash-featurelist.html]: "i" is used twice in this function. I declare it separately avoid redeclaration and give both loops the same form: `for (i = 0; i...`

File: static/elements/chromedash-featurelist.html (hunk 6 above)
Comment (mdittmer) [inline static/elements/chromedash-featurelist.html]: The `var j` is wrong; possibly left over from a different implementation. See my previous note on the non-var-iness.

File: static/elements/chromedash-featurelist.html (hunk 3 above)
Comment (mdittmer) [inline static/elements/chromedash-featurelist.html]: Added a bunch of comments.

File: static/elements/chromedash-featurelist.html [hunk 7]
```diff
@@ -130,38 +178,76 @@
...
+          // Query parts of the form: "name : value".
+          var propertyQueries = [];
+          // Query parts of the form: "name = value", "name < value", etc.
+          var operatorQueries = [];
+          // Other words in the query (not matching forms above).
+          var keywordQueries = [];
+
+          // Accumulate keyword and property queries.
+          var i;
+          for (i = 0; i < reMatches.length; i++) {
```
Comment (ebidel) [inline static/elements/chromedash-featurelist.html]: just use `var i = 0` in the for oop

Comment (ebidel) [inline static/elements/chromedash-featurelist.html]: Just `var i = 0` in the for loop

File: static/elements/chromedash-featurelist.html [hunk 8]
```diff
@@ -130,38 +178,76 @@
...
+            this.filtered = this.features;
           } else {
-            // Only one of the platforms needs to match.
-            // Do a general search over misc fields.
-            // Remove "=" prefix so "in development"/"proposed" version queries
-            // return results.
-            if (val.indexOf('=') == 0) {
-              val = val.substring(1);
+            var results = this.features;
+            for (i = 0; i < filters.length; i++) {
```
Comment (ebidel) [inline static/elements/chromedash-featurelist.html]: Just `var i = 0` in the for loop

File: static/elements/chromedash-featurelist.html (hunk 2 above)
Comment (mdittmer) [inline static/elements/chromedash-featurelist.html]: Fixed.

File: static/elements/chromedash-featurelist.html (hunk 4 above)
Comment (mdittmer) [inline static/elements/chromedash-featurelist.html]: Switched to `function(...) { ... for (var i...) { ... } ... for (var i...) { ... } ... }` (harmless variable redeclaration) as per discussion.

File: static/elements/chromedash-featurelist.html (hunk 6 above)
Comment (mdittmer) [inline static/elements/chromedash-featurelist.html]: This is fixed now.

File: static/elements/chromedash-featurelist.html (hunk 7 above)
Comment (mdittmer) [inline static/elements/chromedash-featurelist.html]: Switched to `function(...) { ... for (var i...) { ... } ... for (var i...) { ... } ... }` (harmless variable redeclaration) as per discussion.

File: static/elements/chromedash-featurelist.html (hunk 8 above)
Comment (mdittmer) [inline static/elements/chromedash-featurelist.html]: Switched to `function(...) { ... for (var i...) { ... } ... for (var i...) { ... } ... }` (harmless variable redeclaration) as per discussion.

**Comment by mdittmer** [comment]
Fixed outstanding hanging "," and var decl issues. @ebidel PTAL.

**Comment by ebidel** [comment]
LGTM. Can you rebase so we can merge?

**Comment by mdittmer** [comment]
Rebased. @ebidel: bd26ae910dde0777596c9582591894d1a124ad64 and e9897db9baaa7186a104c7f1968890251b5451d5 are new. PTAL.

**Comment by ebidel** [comment]
LGTM

**Pull request merged**